fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/FDBRecordStore.java

```diff
@@ -2694,6 +2705,56 @@ public void addRebuildRecordCountsJob(List<CompletableFuture<Void>> work) {
         work.add(future);
     }
 
+    /**
+     * Boundaries are a list of key ranges maintained by each FDB server. This information can be useful for splitting a
+     * large task (e.g. rebuild index for a large record store) into small tasks (e.g.  rebuild index for records in
+     * certain primary key ranges) more evenly so that they can be executed in a parallel fashion efficiently. The
+     * returned boundaries are an estimate from FDB's locality API and may not represent the exact boundary locations
+     * at any database version.
+     *
+     * The boundaries are returned as a list, which is sorted and does not have duplication. The first element of the
+     * list is <code>low</code> and the last element is the <code>high</code>.
+     *
+     * @param low low endpoint of primary key range (inclusive)
+     * @param high high endpoint of primary key range (exclusive)
+     * @return the list of boundaries of primary keys
+     */
+    @Nonnull
+    public CompletableFuture<List<Tuple>> getPrimaryKeyBoundariesAsync(@Nonnull Tuple low, @Nonnull Tuple high) {
+        final Transaction transaction = ensureContextActive();
+        byte[] rangeStart = recordsSubspace().pack(low);
+        byte[] rangeEnd = recordsSubspace().pack(high);
+        CloseableAsyncIterator<byte[]> cursor = LocalityUtil.getBoundaryKeys(transaction, rangeStart, rangeEnd);
+        final List<CompletableFuture<List<KeyValue>>> fdbBoundaryEntriesFutures = new ArrayList<>();
+        return AsyncUtil.whileTrue(() ->
+                cursor.onHasNext().thenApply(hasNext -> {
+                    if (hasNext) {
+                        byte[] result = cursor.next();
+                        fdbBoundaryEntriesFutures.add(transaction.snapshot().getRange(result, rangeEnd, 1).asList());
+                    }
+                    return hasNext;
+                }), context.getExecutor())
+                .thenCompose(vignore -> AsyncUtil.getAll(fdbBoundaryEntriesFutures))
```

Comment:
If I'm reading this correctly, this will perform all of these reads at unlimited parallelism, which maybe is fine.